// @mfelibs/webpack-config/build/webpack.base.conf.js
var path = require('path')
var utils = require('./utils')
var config = require('./config')
var fs = require('fs');
var glob = require('glob');
var cwd = process.cwd();
var marauderConfig = require(path.resolve(cwd,'marauder.config.js'))
function resolve(dir) {
	return path.join(cwd, '..', dir)
}
var ROOT_PATH = process.cwd();
// 获取指定路径下的入口文件
function getEntries(globPath) {
	var files = glob.sync(globPath),
		entries = {};
	files.forEach(function(filepath) {
		// 取倒数第二层(view下面的文件夹)做包名
		var split = filepath.split('/');
		var name = split[split.length - 2];
		entries[name] = './' + filepath;
	});
	return entries;
}
var entries = getEntries('src/view/**/index.js');
var entry = {}
Object.keys(entries).forEach(function(name) {
	// 每个页面生成一个entry，如果需要HotUpdate，在这里修改entry
	entry[name] = entries[name];
})
module.exports = {
	entry: entry,
	output: {
		path: config.build.assetsRoot,
		filename: '[name].js',
		publicPath: process.env.NODE_ENV === 'production' ? config.build.assetsPublicPath : config.dev.assetsPublicPath
	},
	resolve: {
		extensions: ['.js', '.json'],
		modules: [
			resolve('src'),
			resolve('node_modules')
		],
		alias: {
			'src': resolve('src')
		}
	},
	module: {
		rules: [{
			test: /\.js$/,
			loader: 'babel-loader',
			include: [resolve('src'), resolve('test')]
		}, {
			test: /\.(png|jpe?g|gif|svg)(\?.*)?$/,
			loader: 'url-loader',
			query: {
				limit: 2000,
				name: marauderConfig.hash ? utils.assetsPath('img/[name].[hash:7].[ext]') : utils.assetsPath('img/[name].[ext]')
			}
		}, {
			test: /\.(woff2?|eot|ttf|otf)(\?.*)?$/,
			loader: 'url-loader',
			query: {
				limit: 2000,
				name: utils.assetsPath('fonts/[name].[hash:7].[ext]')
			}
		}]
	}
}
